### ISSUE TITLE
Introduce support for annotation attribute overrides within annotation hierarchies

### ISSUE THREAD

**Issue by mibutec** [junit-team/junit5#1440]
## Overview

I'm working on some tests using scenario based testing. To mark those tests I'm using an annotation `@Scenario("Description of my scenario")`. This results in some reports containing the scenario title and some information about execution of that scenario. The scenario title works for my custom reports, but not for JUnit reporting, since `@DisplayName` isn't set.

A possible solution would be to annotate those scenarios with `@DisplayName`, i.e.

```java
@Scenario
@DisplayName("Description of my scenario")
```

However, that's not my favored solution, since there is one more annotation cluttering my code. I would like to be able to inherit annotations including their attributes.

A possible solution could be something like

```java
@Retention(RUNTIME)
@Target(METHOD)
@Test
@InheritedAnnotation(inheritFrom=DisplayName.class)
public @interface Scenario {
	String value();
}
```

This would lead to `@DisplayName` being found as an annotation when `@Scenario` is scanned. Since `@Scenario` declares a `value` attribute with the same type as defined in `@DisplayName`s `value` attribute, the attribute is inherited.

The default behavior for attribute inheritance would be something like: All attributes with same signature are inherited from `sourceAnnotation` to `targetAnnotation`. For all others default implementation is used.

One might think of something like:

```java
@Retention(RUNTIME)
@Target(METHOD)
@Test
@InheritedAnnotation(inheritFrom=DisplayName.class, attributeMapper=MyCustomAttributeMapper.class)
public @interface Scenario {
	String value();
}
```

to allow writing one's own attribute inheritance strategies.

## Related Issues

- #614
- #1504
- #1543

**Comment by sbrannen** [comment]
Hi @mibutec,

What you are proposing is essentially the support for `@AliasFor` that I introduced in the Spring Framework a few years ago (building on Spring's previous support for "annotation attribute overriding" based on naming conventions alone). 

Due to the complexity of the implementation of such a feature, I do not foresee the JUnit Team supporting that any time in the near future.

Note, however, that we actually do think it would be both cool and useful to have such support in JUnit itself. To be perfectly honest, we think it should be a feature in Java itself (or perhaps in a stand-alone third-party library).

In other words, we totally agree that it would be a nice feature to have, but implementing such a feature is currently not on the team's radar.

In light of that, I'll move this issue to the general backlog.

**Comment by sbrannen** [comment]
FYI: I took the liberty to reword the title and description of this issue to align with terminology we typically use within JUnit 5 (which aligns with [terminology used by Spring](https://github.com/spring-projects/spring-framework/wiki/Spring-Annotation-Programming-Model)).

**Comment by mibutec** [comment]
Hi @sbrannen,

if we are only talking about efforts, I could provide a Pull Request. What would be the minimal solution?

1. A solution as described above would be quite fast to implement. It might not be that convenient as `@AliasFor`, but when implemented as repeatable Annotation it would be that powerful.
2. Re-implementing `@AliasFor` would be the next level, seeming to blow up JUnits code quite more.
3. Extracting the `@AliasFor` code to an own library and reusing it would be the even higher level, coming with a lot of complextity (for me) since Spring Project would be effected.

What would be the minimum level for a PR to get that "cool and useful to support" in a PR?

**Comment by sormuras** [comment]
> if we are only talking about efforts, I could provide a Pull Request.

Great! File such a PR against openjdk please. Would really be a great addition to the language.

Why not JUnit?

> Due to the complexity of the implementation of such a feature, I do not foresee the JUnit Team supporting that any time in the near future.

I agree with Sam here.

**Comment by sbrannen** [comment]
> What would be the minimum level for a PR to get that "cool and useful support" in a PR?

For me the bare minimum would be something on par with Spring's `@AliasFor` support, and as I mentioned that is out of scope for JUnit at the moment.

I am against any kind of half-baked solution to be replaced at a later date.

The JUnit 5 Team has actually discussed _something similar to `@AliasFor`_ on several occasions and repeatedly, unanimously decided against implementing it within JUnit.

**Comment by sbrannen** [comment]
Related to: 

- #614
- #1504
- #1543

**Comment by NicklasWallgren** [comment]
Has there been any new discussions around this as of late?
